data/transactions/logic/assembler.go

```diff
@@ -1802,57 +1856,141 @@ func (ops *OpStream) typeErrorf(format string, args ...interface{}) {
 	}
 }
 
+func (ops *OpStream) snapshotStack() {
+	// Snapshot the stack after every call to trackStack
+	snap := make([]StackType, len(ops.known.stack))
+	copy(snap, ops.known.stack[:])
+	ops.stacks = append(ops.stacks, snap)
+}
+
 // trackStack checks that the typeStack has `args` on it, then pushes `returns` to it.
-func (ops *OpStream) trackStack(args StackTypes, returns StackTypes, instruction []string) {
+func (ops *OpStream) trackStack(args StackTypes, returns StackTypes, instruction lineTokens) {
 	// If in deadcode, allow anything. Maybe it's some sort of onchain data.
 	if ops.known.deadcode {
 		return
 	}
-	argcount := len(args)
-	if argcount > len(ops.known.stack) && ops.known.bottom == StackNone {
+
+	expectedCnt, actualCnt := len(args), len(ops.known.stack)
+	if expectedCnt > actualCnt && ops.known.bottom.AVMType == StackNone.AVMType {
 		ops.typeErrorf("%s expects %d stack arguments but stack height is %d",
-			strings.Join(instruction, " "), argcount, len(ops.known.stack))
+			strings.Join(instruction.strings(), " "), expectedCnt, actualCnt)
 	} else {
-		firstPop := true
-		for i := argcount - 1; i >= 0; i-- {
-			argType := args[i]
-			stype := ops.known.pop()
-			if firstPop {
-				firstPop = false
-				ops.trace("pops(%s", argType)
-			} else {
-				ops.trace(", %s", argType)
-			}
-			if !typecheck(argType, stype) {
+		for i := expectedCnt - 1; i >= 0; i-- {
+			expectedType, actualType := args[i], ops.known.pop()
+			if !actualType.AssignableTo(expectedType) {
 				ops.typeErrorf("%s arg %d wanted type %s got %s",
-					strings.Join(instruction, " "), i, argType, stype)
+					strings.Join(instruction.strings(), " "), i,
+					expectedType.String(), actualType.String())
 			}
 		}
-		if !firstPop {
-			ops.trace(")")
-		}
+		ops.trace("pops%s ", args.String())
 	}
 
 	if len(returns) > 0 {
 		ops.known.push(returns...)
-		ops.trace(" pushes(%s", returns[0])
-		if len(returns) > 1 {
-			for _, rt := range returns[1:] {
-				ops.trace(", %s", rt)
-			}
+		ops.trace("pushes%s ", returns.String())
+	}
+}
+
+// assembleTokens works on a single set of tokens representing
+// an op and its immediate args
+func (ops *OpStream) assembleTokens(current []token) {
```

Comment:
I chose to break out this method and `assembleLine` with the thought that we may want to do some prep/cleanup work before and after a call to the guts of the assembler.

Although, initially the work I'd done that benefited from this breakout was to snapshot the stack and I've since removed that feature.

Reply:
I think I'm going to revert this change, it does not make sense to change if there is no benefit from isolating it this way and would make it easier to review the rest